### PR TITLE
wip: Add support for SRIOV

### PR DESCRIPTION
**Pull request by amshinde**
This PR adds support for passing SRIOV supported NICs to the VM. 
We check if the network interface passed in the network namespace is a physical device. If a physical network device is found, we skip the bridge creation flow and pass the device directly to the VM through VFIO passthrough. To do this, we unbind the network device from its network driver and bind it to VFIO driver. We do the opposite while stopping the VM.

### CODE REVIEW

File: cnm.go [hunk 1]
```diff
@@ -136,23 +137,37 @@ func (n *cnm) createEndpointsFromScan(networkNSPath string) ([]Endpoint, error)
...
-		} else {
-			endpoint, err = createNetworkEndpoint(idx, uniqueID, netIface.iface.Name)
-			if err != nil {
-				return []Endpoint{}, err
+		}
+
+		err := doNetNS(networkNSPath, func(_ ns.NetNS) error {
+			// Check if interface is a physical interface. Do not create
+			// tap interface/bridge if it is.
+			isPhysical := checkPhysicalIface(netIface.iface.Name)
```
Comment (jodh-intel) [inline cnm.go]: Might be useful to log `isPhysical` along with the ifname.

**Comment by jodh-intel** [comment]
lgtm.

Aside: I wonder if it's worth having an in-tree-document describing the available features.

**Comment by amshinde** [comment]
@jodh-intel Thanks for taking a look. Yes, I am planning to move the SRIOV documentation that we did for 2.1 here.

File: cnm.go (hunk 1 above)
Comment (amshinde) [reply cnm.go]: Yes, will be useful to log that as well. Will add it.

File: utils.go [hunk 2]
```diff
@@ -78,3 +78,19 @@ func cleanupFds(fds []*os.File, numFds int) {
 		_ = fds[i].Close()
 	}
 }
+
+// writeToFile opens a file in write only mode and writes bytes to it
+func writeToFile(path string, data []byte) error {
+	f, err := os.OpenFile(path, os.O_WRONLY, 0755)
```
Comment (sboeuf) [inline utils.go]: You might want to define a const `fileMode0755` that could be shared across the entire package.

File: hyperstart.go [hunk 3]
```diff
@@ -172,9 +173,13 @@ func (h *hyper) buildNetworkInterfacesAndRoutes(pod Pod) ([]hyperstart.NetworkIf
 	var ifaces []hyperstart.NetworkIface
 	var routes []hyperstart.Route
 	for _, endpoint := range networkNS.Endpoints {
-		netIface, err := getNetIfaceByName(endpoint.GetName(), netIfaces)
-		if err != nil {
-			return []hyperstart.NetworkIface{}, []hyperstart.Route{}, err
+		var netIface net.Interface
```
[the remark this answers is not in the document]
Comment (sboeuf) [reply hyperstart.go]: @amshinde yes we talked about it with @mcastelino. However, this should be done in a follow up PR.

File: network.go [hunk 4]
```diff
@@ -847,6 +972,74 @@ func checkPhysicalIface(ifaceName string) bool {
...
+		return nil, err
+	}
+
+	// Get Driver
+	driver, err := ethHandle.DriverName(ifaceName)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get vendor and device id from pci space (sys/bus/pci/devices/$bdf)
```
[the remark this answers is not in the document]
Comment (amshinde) [reply network.go]: @mcastelino I am in the network namespace when this is called from cnm.go. The device is no longer visible on the host side.

File: cnm.go [hunk 5]
```diff
@@ -136,23 +137,37 @@ func (n *cnm) createEndpointsFromScan(networkNSPath string) ([]Endpoint, error)
...
+			return []Endpoint{}, err
 		}
 
 		routes, err := n.getNetIfaceRoutesWithinNetNs(networkNSPath, netIface.iface.Name)
 		if err != nil {
 			return []Endpoint{}, err
 		}
 
-		endpoint.Properties, err = n.createResult(netIface.iface, netIface.addrs, routes)
+		properties, err := n.createResult(netIface.iface, netIface.addrs, routes)
```
[the remark this answers is not in the document]
Comment (amshinde) [reply cnm.go]: @mcastelino yes, I am getting the MTU for physical endpoints in a later commit of this PR. The code path for virtual endpoints has remained the same. Lets do that in a separate PR.

File: cnm.go [hunk 6]
```diff
@@ -136,23 +137,37 @@ func (n *cnm) createEndpointsFromScan(networkNSPath string) ([]Endpoint, error)
...
+
+		err := doNetNS(networkNSPath, func(_ ns.NetNS) error {
+			// Check if interface is a physical interface. Do not create
+			// tap interface/bridge if it is.
+			isPhysical := checkPhysicalIface(netIface.iface.Name)
+
+			if isPhysical {
+				endpoint, err = createPhysicalEndpoint(netIface.iface.Name)
+			} else {
+				endpoint, err = createVirtualNetworkEndpoint(idx, uniqueID, netIface.iface.Name)
```
[the remark this answers is not in the document]
Comment (amshinde) [reply cnm.go]: For creating a physical endpoint, it is required to be in the network namespace to order to set its attributes.

File: qemu.go [hunk 7]
```diff
@@ -329,22 +329,25 @@ func networkModelToQemuType(model NetInterworkingModel) ciaoQemu.NetDeviceType {
...
-				MACAddress:    endpoint.NetPair.TAPIface.HardAddr,
-				DownScript:    "no",
-				Script:        "no",
-				VHost:         true,
-				DisableModern: q.nestedRun,
-				FDs:           endpoint.NetPair.VMFds,
-			},
-		)
-	}
+		virtualEndpoint, ok := endpoint.(*VirtualEndpoint)
```
[the remark this answers is not in the document]
Comment (amshinde) [reply qemu.go]: @mcastelino We follow the same code as before for virtual endpoints after checking for virtual endpoint. Physical endpoints are attached via a separate code path.

File: cnm.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (amshinde) [reply cnm.go]: I anyways need to be in the network namespace for creating physical endpoint

File: network.go [hunk 8]
```diff
@@ -97,10 +100,121 @@ type NetworkConfig struct {
...
 }
 
-// Endpoint gathers a network pair and its properties.
-type Endpoint struct {
+// Endpoint represents a physical or virtual network interface.
+type Endpoint interface {
+	GetProperties() types.Result
+	GetName() string
+	GetHardwareAddr() string
+	GetType() EndpointType
```
[the remark this answers is not in the document]
Comment (amshinde) [reply network.go]: I used "Get" here  to avoid conflicts as the attributes themselves need to be exported for json handling,

**Comment by sboeuf** [comment]
Waiting for @mcastelino before to merge this

**Comment by sameo** [comment]
@sboeuf I will continue this review later today.

File: qemu.go [hunk 9]
```diff
@@ -327,27 +327,28 @@ func networkModelToQemuType(model NetInterworkingModel) ciaoQemu.NetDeviceType {
...
-					DisableModern: q.nestedRun,
-					FDs:           virtualEndpoint.NetPair.VMFds,
-				},
-			)
-		}
+var networkIndex = 0
 
+func (q *qemu) appendNetwork(devices []ciaoQemu.Device, endpoint Endpoint) []ciaoQemu.Device {
+	virtualEndpoint, ok := endpoint.(*VirtualEndpoint)
+	if ok {
```
Comment (sameo) [inline qemu.go]: Again here we should switch on the interface type for making the code more generic.

File: pkg/ethtool/ethtool.go [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+
+// DriverName returns the driver name of the network interface.
+func DriverName(intf string) (string, error) {
+	e, err := NewEthtool()
+	if err != nil {
+		return "", err
+	}
+	defer e.Close()
+	return e.DriverName(intf)
+}
```
Comment (sameo) [inline pkg/ethtool/ethtool.go]: Please remove the ethtool vendoring and use that package instead.

File: qemu.go [hunk 11]
```diff
@@ -327,22 +327,26 @@ func networkModelToQemuType(model NetInterworkingModel) ciaoQemu.NetDeviceType {
 	}
 }
 
-func (q *qemu) appendNetworks(devices []ciaoQemu.Device, endpoints []Endpoint) []ciaoQemu.Device {
-	for idx, endpoint := range endpoints {
+var networkIndex = 0
+
+func (q *qemu) appendNetwork(devices []ciaoQemu.Device, endpoint Endpoint) []ciaoQemu.Device {
+	virtualEndpoint, ok := endpoint.(*VirtualEndpoint)
```
Comment (sameo) [inline qemu.go]: Again, we should switch on the `Endpoint` type to make the code more generic and understandable.

File: hyperstart.go [hunk 12]
```diff
@@ -172,13 +173,18 @@ func (h *hyper) buildNetworkInterfacesAndRoutes(pod Pod) ([]hyperstart.NetworkIf
 	var ifaces []hyperstart.NetworkIface
 	var routes []hyperstart.Route
 	for _, endpoint := range networkNS.Endpoints {
-		netIface, err := getNetIfaceByName(endpoint.NetPair.VirtIface.Name, netIfaces)
-		if err != nil {
-			return []hyperstart.NetworkIface{}, []hyperstart.Route{}, err
+		var netIface net.Interface
+
+		endpointType := endpoint.GetType()
+		if endpointType == VirtualEndpointType {
```
Comment (sameo) [inline hyperstart.go]: A cleaner way would be to switch on the endpoint:

```Go
switch ep := endpoint.(type) {
     case *VirtualEndpoint:
               if err := xconnectVMNetwork(&(ep.NetPair), true); err != nil {
					return err
				}
}
```

File: network.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (amshinde) [reply network.go]: @sameo What if I change the methods to EndpointProperties()..and so on?

File: network.go [hunk 13]
```diff
@@ -194,14 +308,20 @@ func runNetworkCommon(networkNSPath string, cb func() error) error {
 
 func addNetworkCommon(pod Pod, networkNS *NetworkNamespace) error {
 	err := doNetNS(networkNS.NetNsPath, func(_ ns.NetNS) error {
-		for idx := range networkNS.Endpoints {
-			if err := xconnectVMNetwork(&(networkNS.Endpoints[idx].NetPair), true); err != nil {
-				return err
+		for idx, endpoint := range networkNS.Endpoints {
+			endpointType := endpoint.GetType()
+
+			if endpointType == VirtualEndpointType {
```
[the remark this answers is not in the document]
Comment (amshinde) [reply network.go]: @sameo I have changed invocations at other places. The above code is anyways removed in a later commit when I add Attach and Detach interface methods for plysical and virtual network devices and simply call into those.

**Comment by amshinde** [comment]
@sameo I have pushed changes to introduce switching on the endpoint type. Have left a comment on the interface names. Let me know if you have a preference on it, before I go ahead and make the changes.

File: network.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (sameo) [reply network.go]: @amshinde Sorry to be a pain but this is going to stutter. Idiomatic Go tells us getters should not use the `Get` prefix, so please rename the few conflicting fields.

**Comment by sameo** [comment]
@amshinde Thanks for the switch fixes, I also commented on the Endpoint interface getter. Once this is fixed, I'll approve that PR.

**Comment by sameo** [comment]
LGTM

File: network.go [hunk 14]
```diff
@@ -156,6 +159,29 @@ func (endpoint *VirtualEndpoint) SetProperties(properties types.Result) {
 	endpoint.Properties = properties
 }
 
+func networkLogger() *logrus.Entry {
+	return virtLog.WithField("subsystem", "network")
+}
+
+// Attach for virtual endpoint bridges the network pair and adds the
+// tap interface of the network pair to the hypervisor.
+func (endpoint *VirtualEndpoint) Attach(h hypervisor) error {
```
Comment (mcastelino) [inline network.go]: @amshinde unrelated to this PR. I will rewrite this to have different virtual endpoints with their own Attach and detach methods.

File: hyperstart.go [hunk 15]
```diff
@@ -206,9 +206,11 @@ func (h *hyper) buildNetworkInterfacesAndRoutes(pod Pod) ([]hyperstart.NetworkIf
 			MACAddr:     endpoint.GetHardwareAddr(),
 		}
 
-		switch endpoint.(type) {
+		switch ep := endpoint.(type) {
 		case *VirtualEndpoint:
 			iface.MTU = netIface.MTU
+		case *PhysicalEndpoint:
```
Comment (mcastelino) [inline hyperstart.go]: @amshinde why do you not use an GetMTU() interface here. Then you do not need to switch on the concrete type.

Comment (mcastelino) [reply hyperstart.go]: Also all the more reason for https://github.com/containers/virtcontainers/pull/470

**Comment by sboeuf** [comment]
@mcastelino do you expect more changes or can I merge this PR ?

**Comment by mcastelino** [comment]
@amshinde please update the SRIOV CNM plugin documentation to show that we can use SRIOV with CC 3.0. https://github.com/clearcontainers/sriov

**Pull request merged**